File: rest-entities/src/main/java/org/chronopolis/rest/entities/projections/PartialBag.java
package org.chronopolis.rest.entities.projections;

import com.querydsl.core.annotations.QueryProjection;
import lombok.Getter;
import lombok.ToString;
import org.chronopolis.rest.models.enums.BagStatus;

import java.time.ZonedDateTime;
import java.util.Set;

/**
 * @author shake
 */
@Getter
@ToString
public class PartialBag {
    private final Long id;
    private final Long size;
    private final Long totalFiles;
    private final String name;
    private final String creator;
    private final String depositor;
    private final BagStatus status;
    private final ZonedDateTime createdAt;
    private final ZonedDateTime updatedAt;
    private final Set<String> replicatingNodes;

    @QueryProjection
    public PartialBag(Long id,
                      String name,
                      String creator,
                      Long size,
                      Long totalFiles,
                      BagStatus status,
                      ZonedDateTime createdAt,
                      ZonedDateTime updatedAt,
                      String depositor,
                      Set<String> replicatingNodes) {
        this.id = id;
        this.name = name;
        this.creator = creator;
        this.size = size;
        this.totalFiles = totalFiles;
        this.status = status;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
        this.depositor = depositor;
        this.replicatingNodes = replicatingNodes;
    }
}

File: rest-entities/src/main/kotlin/org/chronopolis/rest/entities/projections/PartialBag.kt
package org.chronopolis.rest.entities.projections

import com.querydsl.core.annotations.QueryProjection
import org.chronopolis.rest.entities.Bag
import org.chronopolis.rest.models.enums.BagStatus
import java.time.ZonedDateTime

/**
 * A partial view of a [Bag], with joins on the Depositor (namespace) and
 * BagDistribution/Node (node.username)
 *
 * @author shake
 */
@Suppress("MemberVisibilityCanBePrivate")
class PartialBag @QueryProjection constructor(
        val id: Long,
        val name: String,
        val creator: String,
        val size: Long,
        val totalFiles: Long,
        val status: BagStatus,
        val createdAt: ZonedDateTime,
        val updatedAt: ZonedDateTime,
        val depositor: String,
        var replicatingNodes: Set<String>) {
    constructor(id: Long,
        name: String,
        creator: String,
        size: Long,
        totalFiles: Long,
        status: BagStatus,
        createdAt: ZonedDateTime,
        updatedAt: ZonedDateTime,
        depositor: String) : this(id, name, creator, size, totalFiles, status, createdAt, updatedAt, depositor, emptySet())

    override fun toString(): String {
        return "PartialBag[id=$id;name=$name;creator=$creator;size=$size;" +
                "totalFiles=$totalFiles;status=$status;createdAt=$createdAt;" +
                "updatedAt=$updatedAt;depositor=$depositor;replicatingNodes=$replicatingNodes]"
    }
}